### PR TITLE
Make it possible to use Halide::Buffer without a runtime linked

### PR DESCRIPTION
**Pull request by abadams**
using the magic of extern_weak, symbols become null rather than
undefined references

### CODE REVIEW

**Comment by steven-johnson** [comment]
Does this have any hope of working on Windows?

**Comment by abadams** [comment]
No, I think windows will ignore the attribute. Now that it's a separate weak version maybe I should ifdef it out if WIN32 is defined?

**Comment by abadams** [comment]
Actually ifdeffing it out sucks because that breaks Halide::Buffer on windows... I need to find a windows solution to this too. Sigh.

**Comment by abadams** [comment]
I have a possible windows solution, but I'll need to test it and figure out a mingw story too. Stay tuned.

File: src/runtime/HalideRuntime.h [hunk 1]
```diff
@@ -389,8 +389,22 @@ extern int halide_device_sync(void *user_context, struct buffer_t *buf);
...
 extern int halide_device_malloc(void *user_context, struct buffer_t *buf, const struct halide_device_interface *device_interface);
 
+/** Free device memory. */
 extern int halide_device_free(void *user_context, struct buffer_t *buf);
 
+/* A weak variant of halide_device_free. If no Halide runtime is
+ * linked, this symbol will be nullptr rather than a linker error,
+ * allowing code to refer to halide_device_free_weak whether or not a
+ * runtime has actually been linked in. This requires a different
+ * mechanism on Windows and other platforms. */
```
Comment (zvookin) [inline src/runtime/HalideRuntime.h]: "...than on other platforms."?

**Comment by abadams** [comment]
ok, I think this approach is reasonable. ptal

**Comment by zvookin** [comment]
LGTM.

**Pull request merged**